packages/common/src/components/CondaHint.ts

```diff
@@ -0,0 +1,300 @@
+
+      const groups = [...curLine.matchAll(re)];
+
+      if (groups && groups.length > 0) {
+        const {
+          pre,
+          packageName,
+          comp1,
+          version1,
+          comp2,
+          version2
+        } = groups[0].groups;
+
+        const matchLengths = [
+          pre,
+          packageName,
+          comp1,
+          version1,
+          comp2,
+          version2
+        ]
+          .map(e => (e !== undefined ? e.length : 0))
+          .reduce(
+            (list, e) =>
+              list.length ? [...list, list[list.length - 1] + e] : [e],
+            [] as number[]
+          );
+
+        const [
+          preEnd,
+          packageEnd,
+          comp1End,
+          version1End,
+          comp2End,
+          version2End
+        ] = matchLengths;
+
+        if (
+          cur.ch >= preEnd &&
+          (!packageEnd || cur.ch <= packageEnd) /* package */
+        ) {
+          const packagePart = packageName
+            ? packageName.slice(0, cur.ch - preEnd)
+            : undefined;
+          list = (
+            await loadPackagesOfChannels(
+              quetzUrl,
+              getChannels(editor.getValue()),
+              packagePart
+            )
+          ).map(p => ({
+            displayText: `${p.name} [${p.channel}] ${p.summary}`,
+            text: `${p.name}`
+          }));
+          callback({
+            list: list,
+            from: CodeMirror.Pos(cur.line, preEnd),
+            to: CodeMirror.Pos(cur.line, packageName ? packageEnd : preEnd)
+          });
+        } else if (
+          cur.ch >= comp1End &&
+          (!version1End || cur.ch <= version1End) /* version1 */
+        ) {
+          list = await loadVersionsOfChannels(
+            quetzUrl,
+            getChannels(editor.getValue()),
+            packageName,
+            version1
+          );
+          callback({
+            list: list,
+            from: CodeMirror.Pos(cur.line, comp1End),
+            to: CodeMirror.Pos(cur.line, version1 ? version1End : comp1End)
+          });
+        } else if (
+          cur.ch >= comp2End &&
+          (!version2End || cur.ch <= version2End) /* version2 */
+        ) {
+          list = await loadVersionsOfChannels(
+            quetzUrl,
+            getChannels(editor.getValue()),
+            packageName,
+            version2
+          );
+          callback({
+            list: list,
+            from: CodeMirror.Pos(cur.line, comp2End),
+            to: CodeMirror.Pos(cur.line, version2 ? version2End : comp2End)
+          });
+        }
+      }
+    }
+  };
+  condaHint.async = true;
+  CodeMirror.registerHelper('hint', 'yaml', condaHint);
+}
+
+export async function fetchSolve(
+  quetzUrl: string,
+  quetzSolverUrl: string,
+  subdir: string,
+  environment_yml: string
+): Promise<string> {
+  const data = {
+    subdir,
```

Comment:
If `fetch`ing from quetz server is gonna be proxied, you can drop the need to bring `subdir` into the frontend.